### PR TITLE
fixes #4393 - rename ancestry label to title

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

**Comment by isratrade** [comment]
The reason for this PR is that adding 'label' on table 'taxonomies' breaks the Katello plugin.

**Comment by isratrade** [comment]
all tests passed!

File: app/models/hostgroup.rb [hunk 1]
```diff
@@ -18,6 +18,7 @@ class Hostgroup < ActiveRecord::Base
   before_save :remove_duplicated_nested_class
 
   alias_attribute :os, :operatingsystem
+  alias_attribute :label, :fullname
```
Comment (isratrade) [inline app/models/hostgroup.rb]: this allows `label` still to be used for hostgroups for getters and setters, but not db calls

File: db/migrate/20130908100439_delete_orphaned_records.rb [hunk 2]
```diff
@@ -22,7 +22,7 @@ def up
     FactValue.where("fact_name_id NOT IN (?) OR host_id NOT IN (?)", FactName.pluck(:id), Host::Base.pluck(:id)).delete_all
     TaxableTaxonomy.where("taxonomy_id NOT IN (?)", Taxonomy.unscoped.pluck(:id)).delete_all
     HostClass.where("host_id NOT IN (?) OR puppetclass_id NOT IN (?)", Host::Base.pluck(:id), Puppetclass.pluck(:id)).delete_all
-    HostgroupClass.where("hostgroup_id NOT IN (?) OR puppetclass_id NOT IN (?)", Hostgroup.pluck(:id), Puppetclass.pluck(:id)).delete_all
+    HostgroupClass.where("hostgroup_id NOT IN (?) OR puppetclass_id NOT IN (?)", Hostgroup.unscoped.pluck(:id), Puppetclass.pluck(:id)).delete_all
```
Comment (isratrade) [inline db/migrate/20130908100439_delete_orphaned_records.rb:26]: `.unscoped` is needed since the default scope orders by `fullname`, but `fullname` is not a db field yet.

File: app/models/hostgroup.rb (hunk 1 above)
Comment (iNecas) [inline app/models/hostgroup.rb]: Is there any benefit of using the label instead fullname directly? I smell some potential issues later…

**Comment by iNecas** [comment]
[test]

**Comment by ghost** [comment]
please: can we have this merged/built ? katello-installer fails due to this.

File: app/models/hostgroup.rb (hunk 1 above)
Comment (iNecas) [inline app/models/hostgroup.rb]: Anyway, due to urgency, we can discuss this later.

Comment (domcleal) [inline app/models/hostgroup.rb]: If this will cause issues, let's address it now.

File: app/models/concerns/hostext/search.rb [hunk 3]
```diff
@@ -21,7 +21,7 @@ module Search
 
       scoped_search :in => :model,       :on => :name,    :complete_value => true, :rename => :model
       scoped_search :in => :hostgroup,   :on => :name,    :complete_value => true, :rename => :hostgroup
-      scoped_search :in => :hostgroup,   :on => :label,   :complete_value => true, :rename => :hostgroup_fullname
+      scoped_search :in => :hostgroup,   :on => :fullname,   :complete_value => true, :rename => :hostgroup_fullname
```
Comment (domcleal) [inline app/models/concerns/hostext/search.rb]: This will break existing bookmarks?

Comment (domcleal) [inline app/models/concerns/hostext/search.rb]: Indentation's a bit funny too.

File: db/migrate/20140219183342_change_labelto_fullname.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class ChangeLabeltoFullname < ActiveRecord::Migration
+  def change
+    rename_column :taxonomies, :label, :fullname
```
Comment (domcleal) [inline db/migrate/20140219183342_change_labelto_fullname.rb]: Is this even going to do the right thing with Katello loaded?

**Comment by domcleal** [comment]
I don't think this is a good outcome really, we shouldn't really be needing to rename database columns because another plugin has modified one of the core tables.  If plugins really have to add columns to Foreman's tables (and can't use associations, which seems preferable to me), then they should be namespaced in the same way that table names themselves or routes are usually prefixed.

"fullname" isn't a good name for the column, we already use this in domains and it's caused confusion (in the translations, we rename it to "description").  It's more like a description or path?

**Comment by bkearney** [comment]
@domcleal yeah.. but the term name was already there :). I do like path..

**Comment by ehelms** [comment]
@domcleal Around the 'what plugins should do if needing to augment a model' - while there is no real way for Foreman to prevent this (due to the nature of Rails engines and migrations), sounds like some best practices or guidelines on how plugins should handle particular integration points needs to be defined and written down.

**Comment by ohadlevy** [comment]
@domcleal path is already an ancestry method, we can use title or just leave it as fullname, with all due respect to translators, i do think the current blocker is more pressing.

**Comment by ohadlevy** [comment]
@ehelms as a role of thumb, based on this experience, I would suggest that every time a pluging extend an existing table, it should prefix the column with the plugin name.

File: db/migrate/20130908100439_delete_orphaned_records.rb (hunk 2 above)
Comment (ohadlevy) [inline db/migrate/20130908100439_delete_orphaned_records.rb:26]: @isratrade is it better to document it in code vs here, it would be hard to find this info when you refactor one year down the road..

**Comment by isratrade** [comment]
re-committed with changing name to `title`

File: app/models/concerns/hostext/search.rb [hunk 5]
```diff
@@ -21,7 +21,8 @@ module Search
 
       scoped_search :in => :model,       :on => :name,    :complete_value => true, :rename => :model
       scoped_search :in => :hostgroup,   :on => :name,    :complete_value => true, :rename => :hostgroup
-      scoped_search :in => :hostgroup,   :on => :label,   :complete_value => true, :rename => :hostgroup_fullname
+      scoped_search :in => :hostgroup,   :on => :title,   :complete_value => false, :rename => :hostgroup_fullname
```
Comment (isratrade) [inline app/models/concerns/hostext/search.rb]: this was kept for legacy purposes to keep search by `hostgroup_fullname` but with complete_value switched to `false`

File: app/models/concerns/nested_ancestry_common.rb [hunk 6]
```diff
@@ -2,32 +2,40 @@ module NestedAncestryCommon
...
+    after_save :update_matchers , :on => :update, :if => Proc.new {|obj| obj.title_changed?}
 
     validates :name, :presence => true, :uniqueness => {:scope => :ancestry, :case_sensitive => false }
-    validates :label, :presence => true, :uniqueness => true
+    validates :title, :presence => true, :uniqueness => true
 
-    scoped_search :on => :label, :complete_value => :true, :default_order => true
+    scoped_search :on => :title, :complete_value => :true, :default_order => true
+    # for legacy purposes, keep search on :label but without autocomplete
+    scoped_search :on => :title, :complete_value => :false, :rename => :label
```
Comment (isratrade) [inline app/models/concerns/nested_ancestry_common.rb]: this was added for legacy purposes to keep search by `label` but with complete_value switched to false

File: db/migrate/20140128123153_add_ancestry_to_taxonomies.rb [hunk 7]
```diff
@@ -14,7 +15,6 @@ def up
 
   def down
     remove_column :taxonomies, :ancestry
-    remove_index :taxonomies, :ancestry
```
Comment (isratrade) [inline db/migrate/20140128123153_add_ancestry_to_taxonomies.rb:17]: This caused error on migration down.  The index is removed when column is removed so it's not necessary to explicitly remove.

**Comment by isratrade** [comment]
[test]

**Comment by iNecas** [comment]
FYI this doesn't solve the Katello blocker completely, as the Katello migration will happen sooner than adding the label column, that is then renamed to title. Would it be possible, instead of adding new migration just updating the existing one 20140128123153_add_ancestry_to_taxonomies.rb?

File: app/helpers/hosts_helper.rb [hunk 8]
```diff
@@ -198,10 +198,10 @@ def overview_fields host
       [_("Puppet Environment"), (link_to(host.environment, hosts_path(:search => "environment = #{host.environment}")) if host.environment)],
       [_("Host Architecture"), (link_to(host.arch, hosts_path(:search => "architecture = #{host.arch}")) if host.arch)],
       [_("Operating System"), (link_to(host.os.to_label, hosts_path(:search => "os_description = #{host.os.description}")) if host.os)],
-      [_("Host group"), (link_to(host.hostgroup, hosts_path(:search => %Q{hostgroup_fullname = "#{host.hostgroup}"})) if host.hostgroup)],
+      [_("Host group"), (link_to(host.hostgroup, hosts_path(:search => %Q{hostgroup_path = "#{host.hostgroup}"})) if host.hostgroup)],
```
Comment (ohadlevy) [inline app/helpers/hosts_helper.rb]: path or title?

File: db/migrate/20140219183342_change_labelto_fullname.rb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+class ChangeLabeltoFullname < ActiveRecord::Migration
+  def change
+    rename_column :taxonomies, :label, :title
```
Comment (ohadlevy) [inline db/migrate/20140219183342_change_labelto_fullname.rb]: this is going to break existing katello... can we detect if katello is installed?

Comment (iNecas) [inline db/migrate/20140219183342_change_labelto_fullname.rb]: The migration with Katello will fail even sooner: https://github.com/theforeman/foreman/pull/1245#issuecomment-35847029 becuase by the time of running the `20140128123153_add_ancestry_to_taxonomies`, Katello already had added the label to the organization.

Comment (isratrade) [inline db/migrate/20140219183342_change_labelto_fullname.rb]: @iNecas, right, if katello was already installed  before commit 1fa008a4ec7d26855b9b67aac3e5f66b0e3e0670 just six days ago, then the migration in that commit to add column 'label' to 'taxonomies' would fail.  @ohadlevy, so I don't see renaming `label` to `title` affecting katello.

**Comment by isratrade** [comment]
re-committed with fix to hosts_helper

File: db/migrate/20140219183342_change_labelto_fullname.rb (hunk 9 above)
Comment (pitr-ch) [inline db/migrate/20140219183342_change_labelto_fullname.rb]: I think the old migration needs to be fixed instead of adding another one to rename the `label` field. The renaming one could stay so it can conditionally deal with renaming of the `label` field if the unchanged original migration already run in a developer's environment or doing nothing when no `label` is present.

Comment (ohadlevy) [inline db/migrate/20140219183342_change_labelto_fullname.rb]: I dont think we should change old migrations, as it has already been published and there are known users.

I would suggest that Katello namespace the label field in the taxonomy table, and we would rename it to title to avoid clashing object names in the system.

Comment (iNecas) [inline db/migrate/20140219183342_change_labelto_fullname.rb]: So are you saying, that this PR is not going to solve the blocker with Foretello not being installable?

**Comment by isratrade** [comment]
I thought (incorrectly) that the katello plugin migrations were run **after** foreman's migrations on a fresh install.  Because of this line,
https://github.com/Katello/katello/blob/master/lib/katello/engine.rb#L22-L23
they are run in sequence according to date timestamp/ file name.

Thus, katello's migration adds `label` to the db before foreman's migration so this migration 
https://github.com/isratrade/foreman/blob/rename_label/db/migrate/20140219183342_change_labelto_fullname.rb#L3
will cause a failure despite another migration after it that renames `label` to `title`

**Comment by isratrade** [comment]
Here are the options how to remedy the situation:
1) change this migration in foreman
https://github.com/theforeman/foreman/blob/develop/db/migrate/20140128123153_add_ancestry_to_taxonomies.rb#L5 (it is usually not advisable to change past migrations and this is just a temporary band-aid fix and make re-occur in the future)
2) change these lines in katello to run katello migrations **after** foreman rather than adding to the db/migrate path
https://github.com/Katello/katello/blob/master/lib/katello/engine.rb#L22-L23
(this solution doesn't require a change in foreman migrations. rename column works if a clash exists)
3) change katello so all fields added to foreman are namespaced
(this solution is probably best long-term but not short-term, as I cannot estimate how involved the changes were be.)

File: db/migrate/20140219183342_change_labelto_fullname.rb (hunk 9 above)
Comment (isratrade) [inline db/migrate/20140219183342_change_labelto_fullname.rb]: @iNecas, unfortunately not. See my comment below.

**Comment by bbuckingham** [comment]
Is this change only in nightly?  If so, are users using nightly for production?  They should not be as nightly should be considered 'unstable' for a production environment. 

Would another alternative be to give users procedures to rollback the bad migration, update their nightly and then re-apply the migration?

Katello could rename the columns as suggested; however, I really don't think that is a great idea.

**Comment by isratrade** [comment]
@bbuckingham, yes adding ancestry and the `label` column to `taxonomies` table has only been changed in nightlies

**Comment by thomasmckay** [comment]
Strongly against renaming columns in katello at this point. (Agree it's best practice for all sides to prefix with their name foreman_, katello_ etc. but that's another discussion.) Please just fix the migration and we can inform nightly users to run rake db:migrate:redo or such?

**Comment by isratrade** [comment]
@domcleal, I re-committed with a separate commit to squash that changes the previous migration so that `label` field is never created by foreman.

**Comment by thomasmckay** [comment]
ack
Applied patch to foreman/develop & katello/master, rake db:drop db:create db:migrate katello:reset
UI and API working from brief test

File: app/models/taxonomy.rb [hunk 10]
```diff
@@ -25,11 +24,7 @@ class Taxonomy < ActiveRecord::Base
   delegate :import_missing_ids, :inherited_ids, :used_and_selected_or_inherited_ids, :selected_or_inherited_ids, :non_inherited_ids,
            :to => :tax_host
 
-  default_scope lambda { order(:label) }
-
-  def to_param
-    "#{id.to_s.parameterize}"
```
Comment (ohadlevy) [inline app/models/taxonomy.rb:31]: why did you drop this? (out of curiosity) same for hostgroups

File: db/migrate/20140128123153_add_ancestry_to_taxonomies.rb [hunk 11]
```diff
@@ -2,20 +2,20 @@ class AddAncestryToTaxonomies < ActiveRecord::Migration
   def up
     add_column :taxonomies, :ancestry, :string
     add_index :taxonomies, :ancestry
-    add_column :taxonomies, :label, :string
+    add_column :taxonomies, :title, :string
```
Comment (ohadlevy) [inline db/migrate/20140128123153_add_ancestry_to_taxonomies.rb]: this would leave the label in the database.

File: app/models/taxonomy.rb (hunk 10 above)
Comment (isratrade) [inline app/models/taxonomy.rb:31]: @ohadlevy, `to_param` was moved to NestedAncestryCommon, so it wasn't dropped.

File: db/migrate/20140128123153_add_ancestry_to_taxonomies.rb (hunk 11 above)
Comment (ohadlevy) [inline db/migrate/20140128123153_add_ancestry_to_taxonomies.rb]: yes, @isratrade how about something like the following:

if schema_migrations include katello migration id
  dont do anything
else
  create it as label
end

and in the new migration, do the same check, and either rename it to title or create it as title

**Comment by isratrade** [comment]
re-committed with checking katello migration was run that added `label`

File: db/migrate/20140219183342_change_label_to_title.rb [hunk 12]
```diff
@@ -0,0 +1,18 @@
+class ChangeLabelToTitle < ActiveRecord::Migration
+  def change
+    # migration 20131120225132 is run by the katello plugin that adds 'label' to taxonomies
+    if ActiveRecord::Migrator.get_all_versions.include?(20131120225132)
+      add_column :taxonomies, :title, :string
+    else
+      rename_column :taxonomies, :label, :title
+      rename_column :hostgroups, :label, :title
```
Comment (ohadlevy) [inline db/migrate/20140219183342_change_label_to_title.rb]: should this be outside the if statement?

Comment (isratrade) [inline db/migrate/20140219183342_change_label_to_title.rb]: @ohadlevy, fixed

File: app/models/hostgroup.rb [hunk 13]
```diff
@@ -89,7 +85,7 @@ def puppetclass_ids
   def inherited_lookup_value key
     ancestors.reverse.each do |hg|
       if(v = LookupValue.where(:lookup_key_id => key.id, :id => hg.lookup_values).first)
-        return v.value, hg.to_label
```
Comment (domcleal) [inline app/models/hostgroup.rb:92]: Why these changes, since to_label now returns title?

Comment (isratrade) [inline app/models/hostgroup.rb:92]: @domcleal, I can change this back to `to_label`.  it's an alias of the same method in the NestedAncestryCommon mixin

File: app/models/concerns/nested_ancestry_common.rb [hunk 14]
```diff
@@ -2,32 +2,40 @@ module NestedAncestryCommon
...
 
-  def get_label
+  def get_title
     return name if ancestry.empty?
     ancestors.map { |a| a.name + '/' }.join + name
   end
+  alias_method :get_label, :get_title
+
+  def to_param
+    "#{id}-#{get_title.parameterize}"
```
Comment (domcleal) [inline app/models/concerns/nested_ancestry_common.rb:37]: Should this be to_title?

**Comment by isratrade** [comment]
@domcleal, @ohadlevy, sent commit to squash per your comments

File: app/models/concerns/nested_ancestry_common.rb [hunk 15]
```diff
@@ -2,32 +2,41 @@ module NestedAncestryCommon
...
-    attr_name :label
+    attr_name :title
   end
 
-  def to_label
-    return label if label
-    get_label
+  # override title getter
+  def title
+    return read_attribute(:title) if read_attribute(:title).present?
```
Comment (ohadlevy) [inline app/models/concerns/nested_ancestry_common.rb]:
``` ruby
read_attribute(:title) || get_title
```

?

Comment (isratrade) [inline app/models/concerns/nested_ancestry_common.rb]: @ohadlevy, I agree.  fixed.

**Comment by mccun934** [comment]
tested this with a RPM based scratch build of Katello and Foreman and it installed fine.  ACK from me

**Comment by ohadlevy** [comment]
merged as 832c092 thanks @isratrade and sorry for the trouble Katello
